File: server/services/brokerAdapters/tradelocker.ts
/**
 * TradeLocker adapter — fetches filled orders via REST API.
 * Credentials: loginId = email, server = broker server string (e.g. "ICMarkets-Live01"),
 * passwordEnc decrypts to { password: "..." }
 * accountType determines demo vs live base URL.
 */
import type { RawBrokerTrade } from '../brokerSyncService';

function baseUrl(accountType: string): string {
  return accountType === 'live'
    ? 'https://live.tradelocker.com/backend-service'
    : 'https://demo.tradelocker.com/backend-service';
}

interface TLAuth {
  accessToken:  string;
  refreshToken: string;
  accounts: Array<{ id: number; accNum: number; currency: string }>;
}

async function authenticate(
  base: string, email: string, password: string, server: string,
): Promise<TLAuth> {
  const res = await fetch(`${base}/auth/jwt/token`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify({ email, password, server }),
  });
  if (!res.ok) throw new Error(`TradeLocker auth failed: ${res.status}`);
  const data = await res.json() as any;
  if (!data.accessToken) throw new Error('TradeLocker: no accessToken in response');
  return {
    accessToken:  data.accessToken,
    refreshToken: data.refreshToken,
    accounts:     (data.accounts ?? []).map((acc: any) => ({
      id:       acc.id,
      accNum:   acc.accNum ?? acc.accountId ?? acc.number ?? acc.id,
      currency: acc.currency ?? 'USD',
    })),
  };
}

async function fetchOrderHistory(
  base: string, token: string,
  accId: number, accNum: number,
  fromMs: number, toMs: number,
): Promise<any[]> {
  const startDate = new Date(fromMs).toISOString().slice(0, 10);
  const endDate   = new Date(toMs).toISOString().slice(0, 10);
  const url = `${base}/trade/accounts/${accId}/${accNum}/ordersHistory`
    + `?startDate=${startDate}&endDate=${endDate}&limit=500&status=Filled`;
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}`, 'env-id': 'tradelocker' } });
  if (!res.ok) throw new Error(`TradeLocker ordersHistory: ${res.status}`);
  const data = await res.json() as any;
  return Array.isArray(data) ? data : (data.d?.orders ?? data.orders ?? []);
}

export async function fetchTradeLockerTrades(
  email:       string,
  password:    string,
  server:      string,
  accountType: string,
  fromMs:      number,
  toMs:        number,
): Promise<RawBrokerTrade[]> {
  const base = baseUrl(accountType);
  const auth = await authenticate(base, email, password, server);
  if (!auth.accounts.length) throw new Error('TradeLocker: no accounts found');

  const trades: RawBrokerTrade[] = [];

  for (const acc of auth.accounts) {
    const rows = await fetchOrderHistory(base, auth.accessToken, acc.id, acc.accNum, fromMs, toMs);
    for (const o of rows) {
      trades.push({
        externalId:  String(o.id ?? o.orderId),
        symbol:      String(o.tradableInstrumentId ?? o.instrument ?? o.symbol),
        direction:   (o.side ?? o.tradeSide ?? '').toLowerCase() === 'buy' ? 'Long' : 'Short',
        lots:        parseFloat(o.qty ?? o.quantity ?? 0),
        openPrice:   parseFloat(o.price ?? o.openPrice ?? 0),
        closePrice:  parseFloat(o.filledPrice ?? o.closePrice ?? o.price ?? 0),
        openTime:    o.createdAt ? new Date(o.createdAt).getTime() : fromMs,
        closeTime:   o.filledAt  ? new Date(o.filledAt).getTime()  : toMs,
        profit:      parseFloat(o.realizedPnl ?? o.pnl ?? o.profit ?? 0),
        commission:  -Math.abs(parseFloat(o.commission ?? o.fee ?? 0)),
        rawData:     o,
      });
    }
  }

  return trades;
}

File: copy_platform/executors/tradelocker.py
"""
TradeLocker executor — places / closes market orders via REST API.
Credentials: { loginId (email), password, server (broker server string), accountType }
"""
import asyncio, logging
import aiohttp
from dataclasses import dataclass

log = logging.getLogger("executor.tradelocker")


@dataclass
class ExecResult:
    ok:          bool
    external_id: str | None = None
    entry_price: float | None = None
    error:       str | None = None


def _base(account_type: str) -> str:
    return ("https://live.tradelocker.com/backend-service"
            if account_type == "live"
            else "https://demo.tradelocker.com/backend-service")


class TradeLockerExecutor:
    def __init__(self, creds: dict, account_type: str):
        self.email       = creds.get("loginId", "")
        self.password    = creds.get("password", "")
        self.tl_server   = creds.get("server", "")
        self.base        = _base(account_type)

    async def _auth(self, session: aiohttp.ClientSession) -> tuple[str, int, int]:
        async with session.post(f"{self.base}/auth/jwt/token", json={
            "email": self.email, "password": self.password, "server": self.tl_server,
        }) as r:
            if r.status != 200:
                raise RuntimeError(f"TradeLocker auth: {r.status}")
            data     = await r.json()
            token    = data.get("accessToken")
            accounts = data.get("accounts", [])
            if not token or not accounts:
                raise RuntimeError("TradeLocker: auth failed")
            acc = accounts[0]
            return token, acc["id"], acc.get("accNum", acc["id"])

    def _headers(self, token: str) -> dict:
        return {"Authorization": f"Bearer {token}", "env-id": "tradelocked"}

    async def open_position(self, symbol: str, action: str,
                            volume_lots: float, sl: float | None,
                            tp: float | None) -> ExecResult:
        try:
            async with aiohttp.ClientSession() as session:
                token, acc_id, acc_num = await self._auth(session)
                body = {
                    "qty":        volume_lots,
                    "instrument": symbol,
                    "validity":   "GTC",
                    "side":       "buy" if action == "BUY" else "sell",
                    "type":       "market",
                    "stopLoss":   sl,
                    "takeProfit": tp,
                }
                async with session.post(
                    f"{self.base}/trade/accounts/{acc_id}/{acc_num}/orders",
                    json=body, headers=self._headers(token),
                ) as r:
                    data = await r.json()
                    if r.status not in (200, 201):
                        raise RuntimeError(f"TradeLocker order: {r.status} {data}")
                    pos_id = str(data.get("d", {}).get("orderId") or data.get("orderId") or "")
                    return ExecResult(ok=True, external_id=pos_id)
        except Exception as e:
            log.error(f"[TradeLocker] open_position failed: {e}")
            return ExecResult(ok=False, error=str(e))

    async def close_position(self, position_id: int,
                             volume_lots: float) -> ExecResult:
        try:
            async with aiohttp.ClientSession() as session:
                token, acc_id, acc_num = await self._auth(session)
                async with session.delete(
                    f"{self.base}/trade/accounts/{acc_id}/{acc_num}/positions/{position_id}",
                    params={"qty": volume_lots},
                    headers=self._headers(token),
                ) as r:
                    if r.status not in (200, 204):
                        raise RuntimeError(f"TradeLocker close: {r.status}")
                    return ExecResult(ok=True, external_id=str(position_id))
        except Exception as e:
            log.error(f"[TradeLocker] close_position failed: {e}")
            return ExecResult(ok=False, error=str(e))

    async def modify_position(self, position_id: int,
                              sl: float | None, tp: float | None) -> ExecResult:
        try:
            async with aiohttp.ClientSession() as session:
                token, acc_id, acc_num = await self._auth(session)
                body: dict = {}
                if sl is not None: body["stopLoss"]   = sl
                if tp is not None: body["takeProfit"] = tp
                async with session.patch(
                    f"{self.base}/trade/accounts/{acc_id}/{acc_num}/positions/{position_id}",
                    json=body, headers=self._headers(token),
                ) as r:
                    return ExecResult(ok=r.status in (200, 204), external_id=str(position_id))
        except Exception as e:
            log.error(f"[TradeLocker] modify_position failed: {e}")
            return ExecResult(ok=False, error=str(e))
